[WIP][FIXED JENKINS-26684] Sanitize arguments when launching processes on Windows

olivergondza opened the pull request:
[JENKINS-26684](https://issues.jenkins-ci.org/browse/JENKINS-26684)

This fix makes `Launcher`s in core platform aware and quote arguments transparently. Do not merge yet as I still struggle with jenkins-test-harness run on windows. Review needed.

olivergondza committed:
[FIXED JENKINS-26684] Sanitize arguments when launching processes on windows

jtnord commented on core/src/main/java/hudson/Launcher.java:
```diff
@@ -800,6 +800,10 @@ public Proc launch(ProcStarter ps) throws IOException {
             	jobCmd[idx] = jobEnv.expand(ps.commands.get(idx));
 
+            if (Functions.isWindows()) {
+                jobCmd = new ArgumentListBuilder(jobCmd).toWindowsCommand().toCommandArray();
```
What happens if this windows box is running the POSIX subsystem - or cygwin?
at first pass it would appear that this could break those installations.

olivergondza replied:
`toWindowsCommand()` delegates to `cmd.exe` so it should not be a problem IMHO, though I know little about Windows or cygwin and I have never heard about the POSIX subsystem for Win. Actually, the issue I reported appears when Cygwin is installed (I do not have machine without cygwin at hand). I have tested the fix successfully on oracle JDK 7 and 8 (with cygwin).

jglick commented:
No opinion except to say that this is a dangerous change. :-/

olivergondza commented:
There seems to be several plugins using following pattern:
```
if (!launcher.isUnix()) {
        args = args.toWindowsCommand();
}
```

I guess it will suffice to turn `toWindowsCommand` into noop when used more than once.

olivergondza committed:
[FIXED JENKINS-26684] Make ArgumentListBuilder#toWindowsCommand idempotent

olivergondza committed:
[FIXED JENKINS-26684] Encapsulate windows command escaping in Launcher

This deprecates ArgumentListBuilder#toWindowsCommand and making it noop. Launcher detects windows and escapes automatically. Overload #toWindowsCommand(boolean) was just unused.

olivergondza commented:
I do not have enough resources to give this the testing it requires. Closing in favour of https://github.com/jenkinsci/jenkins/pull/1567.
